### PR TITLE
Handle input_format_null_as_default for nested types

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Handle `input_format_null_as_default` for nested types

Cc: @Avogar 

*Note: marked as bug fix to run tests on previous versions on CI too*

### CODE REVIEW

File: tests/queries/0_stateless/02573_insert_null_as_default_null_as_empty_tuple.reference [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+((('root')))
+select * from format(JSONEachRow, 'payload Tuple(pull_request Tuple(merged_by Tuple(login Nullable(String))))', '{"payload" : {"pull_request": {"merged_by": {"login": "root"}}}}') settings input_format_null_as_default=1;
+((('root')))
+select * from format(JSONEachRow, 'payload Tuple(pull_request Tuple(merged_by Tuple(login Nullable(String))))', '{"payload" : {}}') settings input_format_null_as_default=0;
+(((NULL)))
+select * from format(JSONEachRow, 'payload Tuple(pull_request Tuple(merged_by Tuple(login Nullable(String))))', '{"payload" : {}}') settings input_format_null_as_default=1;
+(((NULL)))
+select * from format(JSONEachRow, 'payload Tuple(pull_request Tuple(merged_by Tuple(login Nullable(String))))', '{"payload" : {"pull_request": {"merged_by": null}}}') settings input_format_null_as_default=0; -- { serverError INCORRECT_DATA }
+select * from format(JSONEachRow, 'payload Tuple(pull_request Tuple(merged_by Tuple(login Nullable(String))))', '{"payload" : {"pull_request": {"merged_by": null}}}') settings input_format_null_as_default=1;
```
Comment (azat) [inline tests/queries/0_stateless/02573_insert_null_as_default_null_as_empty_tuple.reference:11]: This is the case that does not work before.

**Comment by Avogar** [comment]
I think we should support null as default in all nested types in general. So this also will work:
```
select * from format(JSONEachRow, 'x Tuple(a UInt32, b String)', '{"x" : {"a" : 1, "b" : null}}')
select * from format(JSONEachRow, 'x Array(String)', '{"x" : ["Hello", null]}')
select * from format(JSONEachRow, 'x Map(String, String)', '{"x" : {"a" : "Hello", "b" : null}}')
```

And it can be done in a simple way. For Tuple we should just change this line:
https://github.com/ClickHouse/ClickHouse/blob/9d116e6f5c4710452f301775ab6b9899498c90a3/src/DataTypes/Serializations/SerializationTuple.cpp#L234
To this:
```
if (settings.null_as_default)
    SerializationNullable::deserializeTextJSONImpl(element_column, istr, settings, elems[element_pos]);
else
    elems[element_pos]->deserializeTextJSON(element_column, istr, settings);
```

**Comment by Avogar** [comment]
Will you do it in this PR or should I create a separate PR?

**Comment by azat** [comment]
@Avogar how neat it is, thanks!

>Will you do it in this PR or should I create a separate one?

Sure, already rebasing.

**Review by Avogar — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Comment by azat** [comment]
Test failures does not related, but interesting:

>[Stateless tests (release, wide parts enabled) — fail: 1, passed: 5047, skipped: 6](https://s3.amazonaws.com/clickhouse-test-reports/46725/134e50ddc406f77d1e22c2c0e04ff3faa88f480c/stateless_tests__release__wide_parts_enabled_.html)

Without taking a closer look I would say that #45226 could be related, since it uses mprotect that should split VMAs.

```
Limit on number of mappings (/proc/sys/vm/max_map_count): 65530.
You should increase the limit for vm.max_map_count in /etc/sysctl.conf
2023.02.23 09:14:21.464106 [ 260478 ] {8ff3954e-b98c-4387-9218-0ef17402c980} <Error> TCPHandler: Code: 1001. DB::Exception: std::bad_alloc: std::bad_alloc. (STD_EXCEPTION), Stack trace (when copying this message, always include the lines below):
```

Let's also take a look at previous CI run:

>[Stress test (debug) — Killed by signal (in clickhouse-server.log)](https://s3.amazonaws.com/clickhouse-test-reports/46725/9ab4944b9ed0d0d126688d153477e5bbfd0d0fdd/stress_test__debug_.html)

```
 [ 1521 ] {} <Fatal> BaseDaemon: Code: 445. DB::ErrnoException: Cannot mprotect memory region, errno: 12, strerror: Cannot allocate memory. (CANNOT_MPROTECT), Stack trace (when copying this message, always include the lines below):
```

>[Stress test (debug) — Killed by signal (in clickhouse-server.log)](https://s3.amazonaws.com/clickhouse-test-reports/46725/134e50ddc406f77d1e22c2c0e04ff3faa88f480c/stress_test__debug_.html)

```
clickhouse-server: ../src/Interpreters/Context.cpp:3660: void DB::Context::setCurrentTransaction(DB::MergeTreeTransactionPtr): Assertion `!merge_tree_transaction || !txn' failed.
```

>[Stress test (msan) — Hung check failed, possible deadlock found (see hung_check.log)](https://s3.amazonaws.com/clickhouse-test-reports/46725/134e50ddc406f77d1e22c2c0e04ff3faa88f480c/stress_test__msan_.html)

Hung of `DROP DATABASE`, maybe ENOMEM could be related

```
2023.02.23 11:26:40.478622 [ 1646 ] {d7b7a9e6-57da-4de4-a417-67ca467c3ce5::all_1_1_0_2} <Error> MutatePlainMergeTreeTask: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.88 GiB (attempt to allocate chunk of 1050431 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED) (version 23.2.1.1)
2023.02.23 11:26:40.954242 [ 1646 ] {d7b7a9e6-57da-4de4-a417-67ca467c3ce5::all_1_1_0_2} <Trace> test_whflm99e.ttl (d7b7a9e6-57da-4de4-a417-67ca467c3ce5): Part all_1_1_0_2 in temporary state can be removed without unlocking shared state
2023.02.23 11:26:41.281651 [ 1646 ] {d7b7a9e6-57da-4de4-a417-67ca467c3ce5::all_1_1_0_2} <Error> virtual bool DB::MutatePlainMergeTreeTask::executeStep(): Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.88 GiB (attempt to allocate chunk of 1050431 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.02.23 11:26:41.283498 [ 1646 ] {} <Debug> MemoryTracker: Peak memory usage to apply mutate/merge in d7b7a9e6-57da-4de4-a417-67ca467c3ce5::all_1_1_0_2: 3.00 MiB.
2023.02.23 11:26:45.509054 [ 15556 ] {cd779563-605f-4d1d-9eba-d078ddfb3fe6} <Debug> executeQuery: (from [::1]:58972) (comment: 01070_mutations_with_dependencies.sql) DROP DATABASE IF EXISTS test_whflm99e  (stage: Complete)
2023.02.23 11:26:45.510221 [ 15556 ] {cd779563-605f-4d1d-9eba-d078ddfb3fe6} <Test> TransactionLog: Beginning transaction (60, 49, 27f88074-12b9-4812-941b-59605d02d720) (6985898159906156098)
2023.02.23 11:26:45.510828 [ 15556 ] {cd779563-605f-4d1d-9eba-d078ddfb3fe6} <Trace> ContextAccess (default): Access granted: DROP DATABASE ON test_whflm99e.*
2023.02.23 11:26:45.511687 [ 15556 ] {cd779563-605f-4d1d-9eba-d078ddfb3fe6} <Trace> ContextAccess (default): Access granted: DROP TABLE ON test_whflm99e.ttl
2023.02.23 11:26:45.513897 [ 15556 ] {cd779563-605f-4d1d-9eba-d078ddfb3fe6} <Debug> DatabaseCatalog: Waiting for table d7b7a9e6-57da-4de4-a417-67ca467c3ce5 to be finally dropped
```

>[Stress test (ubsan) — Killed by signal (in clickhouse-server.log)](https://s3.amazonaws.com/clickhouse-test-reports/46725/134e50ddc406f77d1e22c2c0e04ff3faa88f480c/stress_test__ubsan_.html)

Something with the cache:

```
/var/log/clickhouse-server/clickhouse-server.err.log:2023.02.23 13:33:19.861056 [ 2866 ] {2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b} <Fatal> : Logical error: 'Operation `write` can be done only by downloader. (CallerId: None:2866, downloader id: 2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b:2866)'.
/var/log/clickhouse-server/clickhouse-server.err.log:2023.02.23 13:33:19.861134 [ 2092 ] {2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b} <Fatal> : Logical error: 'Operation `reserve` can be done only by downloader. (CallerId: None:2092, downloader id: 2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b:2092)'.
/var/log/clickhouse-server/clickhouse-server.err.log:2023.02.23 13:33:19.861415 [ 2198 ] {2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b} <Fatal> : Logical error: 'Operation `write` can be done only by downloader. (CallerId: None:2198, downloader id: 2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b:2198)'.
/var/log/clickhouse-server/clickhouse-server.err.log:2023.02.23 13:33:19.861574 [ 2275 ] {2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b} <Fatal> : Logical error: 'Operation `write` can be done only by downloader. (CallerId: None:2275, downloader id: 2a0cbaa7-c3a3-4b27-9873-3f3ffc82985b:2275)'.
```

**Pull request merged**